src/query_builder.erl

```diff
@@ -293,3 +310,250 @@ escape_sql_pattern(<<C:8/integer, Rest/binary>>, Acc)
+  ?assertEqual(Expected, Out).
+
+balance_or3_test() ->
+    In = {'or',
+          {'or',
+           {'or',
+            c,
+            c},
+           c},
+          c},
+    Expected = {'or',
+                 {'or' , c, c},
+                 {'or',  c, c}},
+    Out = simplify(In),
+    ?assertEqual(Expected, Out).
+
+balance_or4_test() ->
+    In = {'or',
+          {'or',
+           {'or',
+            {'or',
+             c,
+             c},
+            c},
+           c},
+          c},
+    Expected = {'or',
+                c,
+                {'or',
+                 {'or', c, c},
+                 {'or', c, c}}},
+    Out = simplify(In),
+    ?assertEqual(Expected, Out).
+
+
+balance_or5_test() ->
+    In = {'or',
+          {'or',
+           {'or',
+            {'or',
+             {'or', c, c},
+             c},
+            c},
+           c},
+          c},
+    Expected = {'or',
+                {'or', c, c},
+                {'or',
+                 {'or' , c, c},
+                 {'or', c, c}}},
+    Out = simplify(In),
+    ?assertEqual(Expected, Out).
+
+
+and_simplify_test() ->
+    In = {'and',
+          {'or',
+           {'or',
+            {'or',
+             {'or',
+              {'or', c, c},
+              c},
+             c},
+            c},
+           c},
+          c1},
+    Expected1 = {'and',
+                {'or',
+                 {'or', c, c},
+                 {'or',
+                  {'or' , c, c},
+                  {'or', c, c}}},
+                c1},
+    Expected2 = {'or',
+                 {'or', {'and', c1, c}, {'and', c1, c}},
+                 {'or',
+                  {'or', {'and', c1, c}, {'and', c1, c}},
+                  {'or', {'and', c1, c}, {'and', c1, c}}}},
+    Out1 = simplify(In),
+    ?assertEqual(Expected1, Out1),
+    Out2 = interalize_and(Expected1),
+    ?assertEqual(Expected2, Out2) .
+
+and_simplify2_test() ->
+    In = {'and',
+          {'or',
+           {'or',
+            {'or',
+             {'or',
+              {'or', c, c},
+              c},
+             c},
+            c},
+           c},
+          {'or', c1, c1}},
+    Expected = {'and',
+                {'or',
+                 {'or', c, c},
+                 {'or',
+                  {'or' , c, c},
+                  {'or', c, c}}},
+                {'or', c1, c1}},
+    Out = simplify(In),
+    ?assertEqual(Expected, Out).
+
+and_simplify3_test() ->
+    In = {'and',
+          {'or',
+           {'or',
+            {'or',
+             {'or',
+              {'or', c, c},
+              c},
+             c},
+            c},
+           c},
+          {'and', c1, c1}},
+    Expected1 = {'and',
+                 {'or',
+                  {'or', c, c},
+                  {'or',
+                   {'or' , c, c},
+                   {'or', c, c}}},
+                 {'and', c1, c1}},
+    Expected2 = {'or',
+                 {'or', {'and', {'and', c1, c1}, c}, {'and', {'and', c1, c1}, c}},
```

Comment:
Line 550 is too long:                  {'or', {'and', {'and', c1, c1}, c}, {'and', {'and', c1, c1}, c}},.